core/src/main/java/org/mule/runtime/core/internal/el/DefaultBindingContextBuilder.java

```diff
@@ -21,87 +25,194 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.function.Supplier;
 
 
 public class DefaultBindingContextBuilder implements BindingContext.Builder {
 
-  private Map<String, Supplier<TypedValue>> bindings;
-  private List<ExpressionModule> modules;
+  private BindingContext parent;
+
+  private TypedValue payloadBinding;
+  private TypedValue attributesBinding;
+  private Supplier<TypedValue> varsBinding;
+
+  private final Map<String, Supplier<TypedValue>> bindings = new HashMap<>();
+  private Collection<ExpressionModule> modules = null;
 
   public DefaultBindingContextBuilder() {
-    this.bindings = new HashMap<>();
-    this.modules = new ArrayList<>();
+    this.parent = NULL_BINDING_CONTEXT;
+    // this.bindings = new HashMap<>();
   }
 
   public DefaultBindingContextBuilder(BindingContext bindingContext) {
-    this.bindings = bindingContext.identifiers().stream().collect(toMap(id -> id, id -> () -> bindingContext.lookup(id).get()));
-    this.modules = new ArrayList<>(bindingContext.modules());
+    this.parent = bindingContext;
+
+    payloadBinding = bindingContext.lookup(PAYLOAD).orElse(null);
+    attributesBinding = bindingContext.lookup(ATTRIBUTES).orElse(null);
+    varsBinding = () -> bindingContext.lookup(VARS).orElse(null);
   }
 
   @Override
   public BindingContext.Builder addBinding(String identifier, TypedValue value) {
-    bindings.put(identifier, () -> value);
+    switch (identifier) {
+      case PAYLOAD:
+        payloadBinding = value;
+        break;
+      case ATTRIBUTES:
+        attributesBinding = value;
+        break;
+      case VARS:
+        varsBinding = () -> value;
+        break;
+      default:
+        bindings.put(identifier, () -> value);
+        break;
+    }
     return this;
   }
 
   @Override
   public BindingContext.Builder addBinding(String identifier, Supplier<TypedValue> lazyValue) {
-    bindings.put(identifier, lazyValue);
+    switch (identifier) {
+      case PAYLOAD:
+        payloadBinding = lazyValue.get();
```

Comment:
Why can't this have lazy values too?

Reply:
the lazy values optimization is for those bindings for which a `TypedValue` is built. `payload` and `attributes` are already `TypedValues`, so no need to make them lazy.